### PR TITLE
feat: add dart version datasource

### PR DESCRIPTION
**Pull request by zeshuaro**
## Changes

Add Dart SDK version datasource

## Context

The introduced datasource is a blocker of #19116, see comment for more details: https://github.com/renovatebot/renovate/pull/19116#issuecomment-1328184197

## Documentation

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/dart-version/index.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      homepage: 'https://dart.dev/',
+      sourceUrl: 'https://github.com/dart-lang/sdk',
+      registryUrl,
+      releases: [],
+    };
+    try {
+      for (const channel of this.channels) {
+        const resp = (
+          await this.http.getJson<DartResponse>(
+            `${registryUrl}/storage/v1/b/dart-archive/o?delimiter=%2F&prefix=channels%2F${channel}%2Frelease%2F&alt=json`
```
Comment (zeshuaro) [inline lib/modules/datasource/dart-version/index.ts:40]: This API call is grabbed when visiting [this](https://dart.dev/get-dart/archive) page to download the Dart SDK.

I tried hitting https://storage.googleapis.com/storage/v1/b/dart-archive/o without the query parameters, which returned more detailed results. But the results are paginated, and I'm not sure how to get to the next page.

**Review by viceice — CHANGES_REQUESTED** [review]
reduce the fixture files to contain only a few versions.

**Comment by zeshuaro** [comment]
The failing test is passing locally for me, so I'm not sure why it's not passing on GitHub Actions 😞

File: lib/modules/datasource/dart-version/index.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        return true;
+      })
+      .flatMap((prefix) => {
+        const version = this.getVersionFromPrefix(prefix);
+        return version ? { version, isStable: channel === 'stable' } : [];
+      });
+
+  // Prefix should have a format of "channels/stable/release/2.9.3/"
+  private getVersionFromPrefix = (prefix: string): string | undefined =>
+    prefix.split('/').at(-2);
```
Comment (viceice) [inline lib/modules/datasource/dart-version/index.ts:78]: `at` is not usable on node v14

File: lib/modules/datasource/dart-version/index.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        result.releases = [...result.releases, ...releases];
+      }
+    } catch (err) {
+      this.handleGenericErrors(err);
+    }
+
+    return result.releases.length ? result : null;
+  }
+
+  private getReleasesFromResponse = (
```
Comment (viceice) [inline lib/modules/datasource/dart-version/index.ts:53]: please convert to normal function declaration

File: lib/modules/datasource/dart-version/index.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        }
+        return true;
+      })
+      .flatMap((prefix) => {
+        const version = this.getVersionFromPrefix(prefix);
+        return version ? { version, isStable: channel === 'stable' } : [];
+      });
+
+  // Prefix should have a format of "channels/stable/release/2.9.3/"
+  private getVersionFromPrefix = (prefix: string): string | undefined => {
```
Comment (viceice) [inline lib/modules/datasource/dart-version/index.ts:77]: same here

File: lib/modules/datasource/dart-version/index.ts (hunk 1 above)
Comment (viceice) [reply lib/modules/datasource/dart-version/index.ts:40]: can we change the prefix to only contain the channel name?

Comment (zeshuaro) [reply lib/modules/datasource/dart-version/index.ts:40]: I've tried and sadly it doesn't work

image

image

File: lib/modules/datasource/dart-version/index.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+          (channel !== 'stable' && stableVersionRegex.test(version))
+        ) {
+          return false;
+        }
+        return true;
+      })
+      .flatMap((prefix) => {
+        const version = this.getVersionFromPrefix(prefix);
+        return version ? { version, isStable: channel === 'stable' } : [];
+      });
```
Comment (viceice) [inline lib/modules/datasource/dart-version/index.ts:74]:
```suggestion
      .map(prefix =>  this.getVersionFromPrefix(prefix))
      .filter(is.string)
      .filter((version) => {
        if (
          version === 'latest' ||
          // The API response contains a stable version being released as a non-stable
          // release. So we filter out these releases here.
          (channel !== 'stable' && stableVersionRegex.test(version))
        ) {
          return false;
        }
        return true;
      })
      .map((version) => ({ version, isStable: channel === 'stable' }));
```

this looks much easier and cleaner

**Review by viceice — APPROVED** [review]

**Pull request merged**